docs/madr/decisions/040-meshservice-hostname-vips.md

```diff
@@ -0,0 +1,127 @@
+# Hostnames and VIP management in MeshService
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/6331
+
+## Context and Problem Statement
+
+With introduction of MeshService, we need to have a mechanism to assign VIPs and hostnames to MeshServices.
+Currently, VIPs are managed in a ConfigMap per Mesh. Hostnames are managed using VirtualOutbound.
+
+## Considered Options
+
+* VIPs management
+* Hostname Generator
+* Static Hostname template in CP config
+
+## Decision Outcome
+
+Chosen option: "Hostname Generator" and "VIPs management" since it's the most flexible option.
+
+## Pros and Cons of the Options
+
+### VIPs management
+
+We have to assign a VIP for each MeshService to be able to address it using transparent proxy.
+
+MeshService object contains VIPs in status.
+```yaml
+status: # managed by CP. Not shared cross zone, but synced to global
+  vip:
+    ip: <kube_cluster_ip> # or kuma VIP or Pod IP for Headless Service
+    type: Kubernetes # | Kuma
+```
+
+If MeshService was created based on Kubernetes Service, we can just take ClusterIP, but otherwise we need to assign each MeshService a Kuma VIP.
+Kuma VIP is a non-routable IP used only by Envoy to indicate destination (like a host header, but on L4). We currently do this with 240.0.0.0/4 CIDR.
+
+Kuma VIPs are assigned asynchronously with MeshService creation. Assigning Kuma VIPs as sync operation on MeshService creation would create a race condition on VIP allocation.
+
+VIP assigning process (single-threaded):
+* List MeshServices
+* If a MeshService has no VIP, assign a Kuma VIP. Kubernetes Services would have a VIP already (ClusterIP), so we don’t need to assign an extra VIP.
+* To assign a VIP, we build an IPAM state reserving every address we already assigned and allocate the next address.
+  To solve the problem of “assigning immediately” a VIP of an old service to a new Service we can also store in memory recently removed VIPs for X seconds and set TTL of DNS responses for X seconds.
+  The current value of 30s is too long, especially that the DNS server is run locally in kuma-dp. Leader election should take more than X seconds, so if we were to lose this list, it’s not a problem.
+
+Following this process, we don’t have to store VIPs in any central place like we do currently (VIP ConfigMap for each Mesh).
+We also assign VIPs for ExternalServices, but this can work as long as we consider both ExternalServices and MeshServices for this process.
```

Comment:
Should ExternalServices just use a different CIDR?

Reply:
good idea!